### PR TITLE
Remove auto-created files from `gradio cc publish` and other tweaks

### LINKED ISSUE #7721: gradio cc publish to twine broken on python 3.9
### Describe the bug

gradio cc publish fails on Python 3.9  with the following error:

```
 Uploading distributions to https://upload.pypi.org/legacy/
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/duerr/miniconda3/envs/gradio_customcomponents/lib/python3.9/site-packages/gradio/cli/comma │
│ nds/components/publish.py:196 in _publish                                                        │
│                                                                                                  │
│   193 │   │   │   │   if (not max_version or max_version in p.name)                              │
│   194 │   │   │   ]                                                                              │
│   195 │   │   │   print(f"Uploading files: {','.join(twine_files)}")                             │
│ ❱ 196 │   │   │   twine_upload(twine_settings, twine_files)                                      │
│   197 │   │   except Exception:                                                                  │
│   198 │   │   │   console.print_exception()                                                      │
│   199 │   if upload_demo and not demo_dir:                                                       │
│                                                                                                  │
│ /home/duerr/miniconda3/envs/gradio_customcomponents/lib/python3.9/site-packages/twine/commands/u │
│ pload.py:123 in upload                                                                           │
│                                                                                                  │
│   120 │   repository_url = cast(str, upload_settings.repository_config["repository"])            │
│   121 │   print(f"Uploading distributions to {repository_url}")                                  │
│   122 │                                                                                          │
│ ❱ 123 │   packages_to_upload = [                                                                 │
│   124 │   │   _make_package(filename, signatures, upload_settings) for filename in uploads       │
│   125 │   ]                                                                                      │
│   126                                                                                            │
│                                                                                                  │
│ /home/duerr/miniconda3/envs/gradio_customcomponents/lib/python3.9/site-packages/twine/commands/u │
│ pload.py:124 in <listcomp>                                                                       │
│                                                                                                  │
│   121 │   print(f"Uploading distributions to {repository_url}")                                  │
│   122 │                                                                                          │
│   123 │   packages_to_upload = [                                                                 │
│ ❱ 124 │   │   _make_package(filename, signatures, upload_settings) for filename in uploads       │
│   125 │   ]                                                                                      │
│   126 │                                                                                          │
│   127 │   repository = upload_settings.create_repository()                                       │
│                                                                                                  │
│ /home/duerr/miniconda3/envs/gradio_customcomponents/lib/python3.9/site-packages/twine/commands/u │
│ pload.py:77 in _make_package                                                                     │
│                                                                                                  │
│    74 │   filename: str, signatures: Dict[str, str], upload_settings: settings.Settings          │
│    75 ) -> package_file.PackageFile:                                                             │
│    76 │   """Create and sign a package, based off of filename, signatures and settings."""       │
│ ❱  77 │   package = package_file.PackageFile.from_filename(filename, upload_settings.comment)    │
│    78 │                                                                                          │
│    79 │   signed_name = package.signed_basefilename                                              │
│    80 │   if signed_name in signatures:                                                          │
│                                                                                                  │
│ /home/duerr/miniconda3/envs/gradio_customcomponents/lib/python3.9/site-packages/twine/package.py │
│ :117 in from_filename                                                                            │
│                                                                                                  │
│   114 │   │   ]                                                                                  │
│   115 │   │   if missing_fields:                                                                 │
│   116 │   │   │   supported_metadata = list(pkginfo.distribution.HEADER_ATTRS)                   │
│ ❱ 117 │   │   │   raise exceptions.InvalidDistribution(                                          │
│   118 │   │   │   │   "Metadata is missing required fields: "                                    │
│   119 │   │   │   │   f"{', '.join(missing_fields)}.\n"                                          │
│   120 │   │   │   │   "Make sure the distribution includes the files where those fields "        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
InvalidDistribution: Metadata is missing required fields: Name, Version.
Make sure the distribution includes the files where those fields are specified, and is using a supported Metadata-Version: 1.0, 
1.1, 1.2, 2.0, 2.1, 2.2.
```

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

See above, the exact same package can be published after upgrading to python 3.10

### Screenshot

_No response_

### Logs

_No response_

### System Info

```shell
python 3.9, gradio 4.21.0
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

The `publish` command would create a README.md, Dockerfile, and requirements.txt file on the fly which would introduce conflicts if the developer manually updated those files and didn't want them to be changed.

The solution is to not create any files in the publish command and instead publish the demo and source directories as-is. Some additional tweaks to the CLI are made.

## Summary of Changes
* The `create` command will create a requirements.txt file in the demo directory and `README.md` file used for the deployed space. The requirements.txt is not modified in the `publish` command anymore. This means that `prefer_local` will upload the wheel but not install it in the space. I think that's ok because we should be encouraging people to publish to pypi and modifying the requirements.txt file can be buggy. A local wheel can't be used with the gradio sdk anyways and I think the benefit of the simple gradio sdk outweighs the cost of not being able to do a local upload.
* The `README.md` will include `short_description` as well as the metadata inputted during the create step.
* The `docs` command will preserve the space metadata in the README between builds.
* The deployed space will now use the gradio sdk instead of the docker sdk.
* The .gitignore file will now ignore the created js files in `templates`.
* Fixes weird bug where the `tool.hatch.artifacts` list has a new entry added to it every time you build.

cc @Wauplin  @radames 

Closes: #7973
Closes: #7105 
Closes #7721

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/cli/commands/components/publish.py [hunk 1]
```diff
@@ -103,9 +65,9 @@ def _publish(
         Optional[Path], Option(help="Path to the demo directory.")
     ] = None,
     source_dir: Annotated[
-        Optional[Path],
+        Path,
```
Comment (freddyaboulton) [inline gradio/cli/commands/components/publish.py:68]: The source_dir is now required so we can find the repo's README but we upload_src is false to preserve the same behavior

**Review by pngwn — APPROVED** [review]
This is much better, and looks great. Have tested locally (as much as I can) and it works well and should be good to merge! Thanks @freddyaboulton!

File: gradio/cli/commands/components/publish.py [hunk 2]
```diff
@@ -217,125 +184,81 @@ def _publish(
...
+            repo_id = f"{whoami['name']}/{package_name}"
+            if not api.repo_exists(repo_id=repo_id, repo_type="space"):
+                api.create_repo(
+                    repo_id=package_name,
+                    repo_type="space",
+                    exist_ok=True,
+                    private=False,
+                    space_sdk="gradio",
+                    token=hf_token,
+                )
```
Comment (Wauplin) [inline gradio/cli/commands/components/publish.py:243]: This snippet is valid but a few things to know:
- no need to check `whoami` or `repo_exists` as it can be implicit in the repo creation step
- it's best to pass the token once and for all when initializing `HfApi` rather than passing it to each individual calls
- it's best not to pass `private=False` as this is already the default behavior. In the future, we might provide an enterprise setting to set new repos to private by default. However if `private=False` is already the default in Gradio, we will not be able to overwrite it (in `huggingface_hub` the default is `private=None`).

So this snippet can be updated to:
 
```suggestion
            api = HfApi(token=hf_token)
            repo_url = api.create_repo(
                repo_id=package_name,
                repo_type="space",
                exist_ok=True,
                space_sdk="gradio",
            )
            repo_id = repo_url.repo_id
```

(+ remove all `token=hf_token` args below)

Comment (Wauplin) [reply gradio/cli/commands/components/publish.py:243]: Or if there's a reason to do the `whoami` + `repo_exists`, would be good to add a comment explaining why (for future ourselves)

File: gradio/cli/commands/components/publish.py [hunk 3]
```diff
@@ -217,125 +184,81 @@ def _publish(
...
-                name = pyproject_toml["project"]["name"]  # type: ignore
-                description = pyproject_toml["project"]["description"]  # type: ignore
 
-            readme_text = README_CONTENTS.format(
-                package_name=package_name,
-                version=version,
-                color_from=color_from,
-                color_to=color_to,
-                template=template,
+            shutil.copyfile(
```
Comment (Wauplin) [inline gradio/cli/commands/components/publish.py:228]: (not related to this line-change)

Inside `with tempfile.TemporaryDirectory() as tempdir:` context, it would be good to create a `.gitignore` file with common rules. Otherwise the uploaded Space contains a bunch of useless files (`__pycache__`, `.ruff_cache`, `.mypy_cache`, maybe some front-end build?, etc.).

FYI, `huggingface_hub`'s `upload_folder` respects the gitignore file at the root of the directory but not all gitignores in subdirectories. This is an `hfh`'s limitation I'm aware of but not planned to be fixed (yet?).  That's why you would need to add more rules to the root gitignore files to avoid folders like `.ruff_cache`.

**Review by Wauplin — COMMENTED** [review]
Thanks for the ping @freddyaboulton! Just to be sure, with this PR the uploaded Space will still be different than the local folder right? Is this something that will be tackled in the future or not really?

Left a couple of comments related to `huggingface_hub` usage, leaving the gradio-specific review to others :D

**Comment by freddyaboulton** [comment]
@Wauplin , with this PR the `demo` folder in the custom component directory will be uploaded to spaces as is. The README, requirements.txt will not be modified and no other files (like a Dockerfile) will be created. The only exception is if `upload_source` if True, the source will be uploaded to the space. This is to not break the `upload_source` behavior as some authors may choose to not upload source.

File: gradio/cli/commands/components/publish.py (hunk 3 above)
Comment (freddyaboulton) [reply gradio/cli/commands/components/publish.py:228]: We create a gitignore file in the `create` step. I will add some more common cases like `mypy_cache` and `ruff_cache` and also copy this file to the `tempdir` used here.

File: gradio/cli/commands/components/publish.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/cli/commands/components/publish.py:243]: Thanks @Wauplin for the suggestion. I misunderstood how `create_repo` works. This suggestion is much better.

**Review by Wauplin — APPROVED** [review]
Nice, thanks for the clarification @freddyaboulton!

> The only exception is if upload_source if True, the source will be uploaded to the space. This is to not break the upload_source behavior as some authors may choose to not upload source.

Do we have an idea of how much `upload_source=False` is really used (and on purpose)? I feel that it's quite a niche use case, especially in an open-source ecosystem. Asking because if this is what makes it impossible to have the exact same git repo for the Space demo and the local setup, then I would consider dropping it. A cool thing about having the same folder structure (not only demo/, but also the source code) would be to allow for contributions from external users. At the moment I feel that publishing the source code of the component on the Hub is only meant for read-only purposes. If someone wants to change something, it has to be applied manually to the local changes by the component author.
(just giving my thoughts here, it's probably out of scope of this PR anyway)

**Comment by pngwn** [comment]
> Do we have an idea of how much upload_source=False is really used (and on purpose)? I feel that it's quite a niche use case, especially in an open-source ecosystem.

tldr: This param exists because GitHub exists and it good at what it does. Spaces is not good at what GitHub is good at.

**More detailed version:**

I feel like I should respond to this because I've been pretty vocal about this in general (and I always use this param, intentionally, and still make my source code available).

Obviously, this param is valuable if you don't want the source to be public but it is always valuable if you want to host your source somewhere else.

I think that spaces is unique as a gitops tool, and the hub in general is exceptional for hosting + maintaining models and datasets. But it is a poor place to maintain software. I don't think this is a goal today nor do i think it will become one in the near future. Spaces (and the hub in general) simple lacks the tooling to be a great place to maintain software, and also complicates workflows + fragments an individuals OSS presence.

I want to maintain all of my OSS work in one place. That place is GitHub. It is the place i have a good workflow. It is the place i get all of my notifications. It is easier for me to track everything I'm working on (which spans languages/ ecosystems/ target audiences), I can't be maintaining software in a ifferent place for each ecosystem. I also think _everyone_ is on GH whereas HF hub is where a subset of people go for specific purposes.

Secondly, I feel strongly about CI. GitHub has a good CI platform, Spaces has none. This makes spaces a non-starter for me personally for maintaining software.

**Comment by Wauplin** [comment]
All clear! Thanks @pngwn for the explanation and point of view. That makes a lot more sense indeed.

I do think that if the correct workflow is:
1. host your component source code on Github
2. host your demo on Spaces

then it should be made clearer in the dev process. For now, I only understood the `upload_source` parameter as "if you don't want to share your source, then use that" (even though source code is rarely hidden with Python). I'm perfectly fine with the workflow above and will definitely adopt it now for the reasons you've mentioned; it just never appeared to me it was the recommended way. 

(next level would be to provide github workflows + guidelines on how to host a custom component project :smile:)

**Comment by pngwn** [comment]
> I do think that if the correct workflow is:
>
> - host your component source code on Github
> - host your demo on Spaces

I don't know if I'd say it is 'correct', we do want to encourage hub usage for some use cases. It is certainly my preferred workflow though, and probably for others too. So maybe explicitly supporting/ documenting that workflow somehow would make sense.

> next level would be to provide github workflows + guidelines on how to host a custom component project

@freddyaboulton and I spoke about this last week and is something i'm thinking about (we could even generate the GH CI files). Pondering a custom action to make it a one-liner too. Will open a thread at some point when i have more thoughts.

**Comment by freddyaboulton** [comment]
Thanks @Wauplin and @pngwn for the discussion! Certainly don't hesitate to use `publish` with the default arguments as that would upload the source @Wauplin . We just wanted to give developers the flexibility to not do so. 

Just wanted to say that there is another reason the space repo structure would not match the local repo. We want to use the gradio image sdk to not have to add a `Dockerfile` and because the gradio sdk feels more appropriate for this use case and is easier to get started with. However, custom requirements must be placed at the top level in a `requirements.txt` and this can conflict with a `requirements.txt` used to develop the component. I think this should be configurable in the Readme metadata but for the time being the repo structure will be slightly different.
